File: 338.counting-bits.cpp
/*
 * @lc app=leetcode id=338 lang=cpp
 *
 * [338] Counting Bits
 */

#include <bits/stdc++.h>
#include "lassert/lassert.h"

using namespace std;

// @lc code=start
class Solution {
 public:
  vector<int> countBits(int n) {
    vector<int> ret;
    for (int i = 0; i <= n; i++) {
      ret.push_back(__builtin_popcount(i));
    }
    return ret;
  }
};
// @lc code=end

int main() {
  Solution s;
  std::vector<int> nums;
  nums = {0, 1, 1};
  ASSERT_EQ(s.countBits(2), nums);
  nums = {0, 1, 1, 2, 1, 2};
  ASSERT_EQ(s.countBits(5), nums);
  nums = {0, 1, 1, 2, 1, 2, 2, 3, 1, 2, 2, 3, 2, 3, 3, 4,
          1, 2, 2, 3, 2, 3, 3, 4, 2, 3, 3, 4, 3, 4, 4, 5};
  ASSERT_EQ(s.countBits(31), nums);
  return 0;
}

File: 338.counting-bits.rs
/*
 * @lc app=leetcode id=338 lang=rust
 *
 * [338] Counting Bits
 */

struct Solution;

#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
    pub right: Option<std::rc::Rc<std::cell::RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

#[macro_export]
macro_rules! list_node {
    () => {
        None
    };
    ( $x:expr,$( $y:expr ),* ) => {
        {
            let mut head = Some(Box::new(ListNode::new($x)));
            let mut tail = &mut head;
            $(
                let node = Some(Box::new(ListNode::new($y)));
                tail.as_mut().unwrap().next = node;
                tail = &mut tail.as_mut().unwrap().next;
            )*
            head
        }
    };
}

// @lc code=start
impl Solution {
    pub fn count_bits(n: i32) -> Vec<i32> {
        (0..=n).map(|i| i.count_ones() as i32).collect::<Vec<i32>>()
    }
}
// @lc code=end

#[test]
fn test1() {
    assert_eq!(Solution::count_bits(2), vec![0, 1, 1]);
}

#[test]
fn test2() {
    assert_eq!(Solution::count_bits(5), vec![0, 1, 1, 2, 1, 2]);
}
